File: packages/shared/src/types/note.ts
/** Believe Notes. `content` is a TipTap/ProseMirror JSON document — the
 * frontend editor's own document format, opaque to everything except the
 * editor itself and the backend's plain-text extraction for AI/embedding. */
export interface Note {
  id: string;
  title: string;
  content: Record<string, unknown>;
  folderId: string | null;
  tags: string[];
  pinned: boolean;
  archived: boolean;
  /** Opaque display metadata linking this note to another Believe.ai entity
   * (a campaign, interview room, contact, job, design project, ...). Never
   * inferred — only set when the frontend had a real entity in view. */
  linkedEntityType: string | null;
  linkedEntityId: string | null;
  linkedEntityLabel: string | null;
  createdAt: string;
  updatedAt: string;
}

/** List-view shape — omits the full editor content, which can be large, but
 * includes a short plainText preview so the list can show a snippet. */
export interface NoteSummary {
  id: string;
  title: string;
  preview: string;
  folderId: string | null;
  tags: string[];
  pinned: boolean;
  archived: boolean;
  linkedEntityType: string | null;
  linkedEntityId: string | null;
  linkedEntityLabel: string | null;
  updatedAt: string;
}

/** "active" (default) excludes archived and trashed notes; "archived" is only
 * archived notes; "trash" is only soft-deleted notes. */
export type NoteView = "active" | "archived" | "trash";

export interface CreateNoteInput {
  title?: string;
  content?: Record<string, unknown>;
  folderId?: string | null;
  tags?: string[];
  linkedEntityType?: string | null;
  linkedEntityId?: string | null;
  linkedEntityLabel?: string | null;
}

export interface UpdateNoteInput {
  title?: string;
  content?: Record<string, unknown>;
  folderId?: string | null;
  tags?: string[];
  pinned?: boolean;
  archived?: boolean;
  linkedEntityType?: string | null;
  linkedEntityId?: string | null;
  linkedEntityLabel?: string | null;
}

export interface NoteFolder {
  id: string;
  name: string;
  createdAt: string;
}

export interface NoteAttachment {
  id: string;
  noteId: string;
  filename: string;
  contentType: string;
  createdAt: string;
}

export interface RelatedNote {
  id: string;
  title: string;
  score: number | null;
}

export interface NoteSearchResult {
  id: string;
  title: string;
  snippet: string;
  score: number | null;
}

export type NoteTransformAction = "explain" | "simplify" | "summarize" | "fix_grammar" | "generate_example" | "extract_action_items";

export interface NoteTextTransformResult {
  result: string;
}

export interface NoteQuizQuestion {
  question: string;
  options: string[];
  correctIndex: number;
  explanation: string | null;
}

export interface NoteQuizResult {
  questions: NoteQuizQuestion[];
}

export interface NoteFlashcard {
  front: string;
  back: string;
}

export interface NoteFlashcardsResult {
  cards: NoteFlashcard[];
}

export interface NoteChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface NoteTutorResult {
  answer: string;
}

export interface NoteTranscriptCleanupResult {
  title: string;
  markdown: string;
}

export type VoiceCommandType = "create_note" | "add_section" | "summarize" | "generate_questions" | "none";

export interface VoiceCommandResult {
  commandType: VoiceCommandType;
  noteTitle: string | null;
  sectionTitle: string | null;
  questionCount: number | null;
}

/** A saved, spaced-repetition-scheduled flashcard — distinct from the ephemeral
 * cards the quick-look "Flashcards" modal generates but never persists. */
export interface SavedFlashcard {
  id: string;
  noteId: string;
  noteTitle: string;
  front: string;
  back: string;
  dueAt: string;
  repetitions: number;
}

export type ReviewGrade = "again" | "hard" | "good" | "easy";

File: apps/backend/schemas/note.py
"""API-facing shapes for Believe Notes."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class NoteDto(BaseModel):
    id: str
    title: str
    content: dict[str, Any]
    folderId: str | None
    tags: list[str]
    pinned: bool
    archived: bool
    linkedEntityType: str | None
    linkedEntityId: str | None
    linkedEntityLabel: str | None
    createdAt: str
    updatedAt: str


class NoteSummaryDto(BaseModel):
    """List-view shape — omits the full editor content, which can be large,
    but includes a short plainText preview so the list can show a snippet."""

    id: str
    title: str
    preview: str
    folderId: str | None
    tags: list[str]
    pinned: bool
    archived: bool
    linkedEntityType: str | None
    linkedEntityId: str | None
    linkedEntityLabel: str | None
    updatedAt: str


class CreateNoteInput(BaseModel):
    title: str = Field(default="Untitled note")
    content: dict[str, Any] = Field(default_factory=dict)
    folderId: str | None = None
    tags: list[str] = Field(default_factory=list)
    linkedEntityType: str | None = None
    linkedEntityId: str | None = None
    linkedEntityLabel: str | None = None


class UpdateNoteInput(BaseModel):
    title: str | None = None
    content: dict[str, Any] | None = None
    folderId: str | None = None
    tags: list[str] | None = None
    pinned: bool | None = None
    archived: bool | None = None
    linkedEntityType: str | None = None
    linkedEntityId: str | None = None
    linkedEntityLabel: str | None = None


NoteView = Literal["active", "archived", "trash"]


class NoteFolderDto(BaseModel):
    id: str
    name: str
    createdAt: str


class CreateNoteFolderInput(BaseModel):
    name: str = Field(min_length=1)


class NoteAttachmentDto(BaseModel):
    id: str
    noteId: str
    filename: str
    contentType: str
    createdAt: str


class RelatedNoteDto(BaseModel):
    id: str
    title: str
    score: float | None


class NoteSearchResultDto(BaseModel):
    id: str
    title: str
    snippet: str
    score: float | None


class NoteFlashcardDto(BaseModel):
    id: str
    noteId: str
    noteTitle: str
    front: str
    back: str
    dueAt: str
    repetitions: int


ReviewGrade = Literal["again", "hard", "good", "easy"]


class ReviewFlashcardInput(BaseModel):
    grade: ReviewGrade


class SaveFlashcardsInput(BaseModel):
    cardCount: int = Field(default=8, ge=1, le=20)
